torchvision/ops/dice_loss.py

```diff
@@ -0,0 +1,69 @@
+import torch
+import torch.nn.functional as F
+
+from ..utils import _log_api_usage_once
+
+
+def dice_loss(inputs: torch.Tensor, targets: torch.Tensor, reduction: str = "none", eps: float = 1e-8) -> torch.Tensor:
+    r"""Criterion that computes Sørensen-Dice Coefficient loss.
+
+    We compute the Sørensen-Dice Coefficient as follows:
+
+    .. math::
+
+        \text{Dice}(x, class) = \frac{2 |X \cap Y|}{|X| + |Y|}
+
+    Where:
+       - :math:`X` expects to be the scores of each class.
+       - :math:`Y` expects to be thess tensor with the class labels.
+
+    the loss, is finally computed as:
+
+    .. math::
+
+        \text{loss}(x, class) = 1 - \text{Dice}(x, class)
+
+    Args:
+        inputs: (Tensor): A float tensor of arbitrary shape.
+                The predictions for each example.
+        targets: (Tensor): A float tensor with the same shape as inputs. Stores the binary
+                classification label for each element in inputs
+                (0 for the negative class and 1 for the positive class).
+        eps: (float, optional): Scalar to enforce numerical stabiliy.
+        reduction (string, optional): ``'none'`` | ``'mean'`` | ``'sum'``
+                ``'none'``: No reduction will be applied to the output.
+                ``'mean'``: The output will be averaged.
+                ``'sum'``: The output will be summed. Default: ``'none'``.
+
+    Return:
+        Tensor: Loss tensor with the reduction option applied.
+    """
+
+    if not torch.jit.is_scripting() and not torch.jit.is_tracing():
+        _log_api_usage_once(dice_loss)
+
+    # compute softmax over the classes axis
+    p = F.softmax(inputs, dim=1)
+
+    # compute the actual dice score
+    dims = (1, 2, 3)
```

Comment:
I agree with @oke-aditya point that this part of the implementation forces specific size of dims. Hence we can't claim arbitrary input size is supported. Shall we refactor?

Reply:
Thoughts on this? This is similar to what PaddleSeg supports.
```
    # compute softmax over the classes axis
    p = F.softmax(inputs, dim=-1)
    p = p.reshape(p.shape[0], -1)
    targets = F.one_hot(targets) #Or we could directly have one-hot tensor as input
    targets = targets.reshape(p.shape[0], -1)

    intersection = torch.sum(p * targets, dim=-1)
    cardinality = torch.sum(p + targets, dim=-1)

    dice_score = 2.0 * intersection / (cardinality + eps)

    loss = 1.0 - dice_score
```

Reply:
This looks fine, but I would prefer to leave one_hot step to the user. We should take one one hot tensors as input.
Also if we are adopting the implementation from Kornia, we should probably cite them as a comment in our code.

Reply:
`p = p.reshape(p.shape[0], -1)` can also be more idiomatically written as `p = p.flatten(start_dim = 1)`

Reply:
Although I am more used to the first idiom, no preferences.